File: Solutions/python/meetup/meetup.py
from datetime import date, timedelta
import calendar


# subclassing the built-in ValueError to create MeetupDayException
class MeetupDayException(ValueError):
    """Exception raised when the Meetup weekday and count do not result in a valid date.

    message: explanation of the error.

    """

    def __init__(self, message="That day does not exist."):
        self.message = message
        super().__init__(self.message)


def meetup(year, month, week, day_of_week):
    days_of_week = {
        "Monday": 0,
        "Tuesday": 1,
        "Wednesday": 2,
        "Thursday": 3,
        "Friday": 4,
        "Saturday": 5,
        "Sunday": 6,
    }

    if week == "teenth":
        for day in range(13, 20):
            if date(year, month, day).weekday() == days_of_week[day_of_week]:
                return date(year, month, day)
    elif week in ["first", "second", "third", "fourth", "fifth", "last"]:
        _, days_in_month = calendar.monthrange(year, month)

        matched_days = [
            day
            for day in range(1, days_in_month + 1)
            if date(year, month, day).weekday() == days_of_week[day_of_week]
        ]

        if week == "first":
            return date(year, month, matched_days[0])
        elif week == "second":
            return date(year, month, matched_days[1])
        elif week == "third":
            return date(year, month, matched_days[2])
        elif week == "fourth":
            return date(year, month, matched_days[3])
        elif week == "fifth":
            if len(matched_days) < 5:
                raise MeetupDayException()
            return date(year, month, matched_days[4])
        elif week == "last":
            return date(year, month, matched_days[-1])

    raise MeetupDayException("Invalid week or day combination.")

File: Solutions/elixir/meetup/lib/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: Date.t()
  def meetup(year, month, weekday, schedule) do
    first_day = Date.new!(year, month, 1)
    last_day = Date.end_of_month(first_day)

    candidates =
      Date.range(first_day, last_day)
      |> Enum.filter(&(Date.day_of_week(&1) == day_of_week(weekday)))

    case schedule do
      :first -> Enum.at(candidates, 0)
      :second -> Enum.at(candidates, 1)
      :third -> Enum.at(candidates, 2)
      :fourth -> Enum.at(candidates, 3)
      :teenth -> Enum.find(candidates, &(&1.day >= 13 and &1.day <= 19))
      :last -> List.last(candidates)
    end
  end

  defp day_of_week(weekday) do
    case weekday do
      :monday -> 1
      :tuesday -> 2
      :wednesday -> 3
      :thursday -> 4
      :friday -> 5
      :saturday -> 6
      :sunday -> 7
    end
  end
end
